src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -131,37 +138,47 @@ impl<L> Controller<L> {
             in_flight: inner.in_flight as u64
         });
 
-        let current_rtt = inner.current_rtt.update(rtt);
+        if use_rtt {
+            inner.current_rtt.update(rtt);
+        }
+        let current_rtt = inner.current_rtt.average();
         let past_rtt = inner.past_rtt.average();
 
+        // No past measurements, set up initial values.
         if past_rtt == 0.0 {
-            // No past measurements, set up initial values.
-            inner.past_rtt.update(current_rtt);
-            inner.next_update = now + Duration::from_secs_f64(current_rtt);
+            if let Some(current_rtt) = current_rtt {
+                inner.past_rtt.update(current_rtt);
+                inner.next_update = now + Duration::from_secs_f64(current_rtt);
+            }
         } else if past_rtt > 0.0 && now >= inner.next_update {
             #[cfg(test)]
             {
-                stats.averaged_rtt.add(current_rtt, now);
+                if let Some(current_rtt) = current_rtt {
+                    stats.averaged_rtt.add(current_rtt, now);
+                }
                 stats.concurrency_limit.add(inner.current_limit, now);
```

Comment:
Is this correct behavior? If we had a non-rtt response - should we include it into the stats even?

If not - it seems like we can turn off this whole codepath - starting from the `if` starting the chain and up until `if self.in_flight_limit.is_none() {`... Or maybe even more than that.

Reply:
It is possible for there to be valid data in the `current_rtt` average despite the most recent response should not be used for RTT calculation. Consider if we receive 4 acknowledgements and one error before `inner.next_update` is reached. Then `current_rtt` will contain the average of those 4, even though the most recent was an error, and we should use that data for this interval.